core/federated/federate.c

```diff
@@ -1932,8 +1933,26 @@ bool update_max_level(tag_t tag, bool is_provisional) {
         // Safe to complete the current tag
         return (prev_max_level_allowed_to_advance != max_level_allowed_to_advance);
     }
-    for (int i = 0; i < _lf_zero_delay_action_table_size; i++) {
-        lf_action_base_t* input_port_action = _lf_zero_delay_action_table[i];
+#ifdef FEDERATED_DECENTRALIZED
+    size_t action_table_size = _lf_action_table_size;
+    lf_action_base_t** action_table = _lf_action_table;
+#else
+    size_t action_table_size = _lf_zero_delay_action_table_size;
+    lf_action_base_t** action_table = _lf_zero_delay_action_table;
+#endif // FEDERATED_DECENTRALIZED
+    for (int i = 0; i < action_table_size; i++) {
+        lf_action_base_t* input_port_action = action_table[i];
+#ifdef FEDERATED_DECENTRALIZED
+        if (
+            (_lf_action_delay_table[i] == 0 && env->current_tag.time == start_time && env->current_tag.microstep == 0)
```

Comment:
> Is this failing to make a distinction between 0 delay and no delay? I'm guessing it's OK, that this means 0 delay. But just checking.

`_lf_action_delay_table` is meant to include information only for connections that have delay strictly greater than (0, 0), so this includes delays of (0, 1) but not delays of (0, 0). So it does distinguish.

> It seems to me that with a 0 delay (not no delay), lf_delay_strict would return a negative time

This is not the case, and in fact if it were the case then we would have deadlocks -- this is why I reverted the change that you suggested to the PR that added `lf_delay_strict`. This was the upshot of [this commit](https://github.com/lf-lang/reactor-c/pull/269/commits/9dfb1505296b0e5cf0f80f2d3502bfb991dff0ea).

> Also, a comment would help for this whole if statement. I'm guessing something like this:

Yes, that is exactly right.

> If this is right, however, wouldn't this also apply to centralized coordination?

Yes, but IIRC it is more than we really need because in centralized coordination, the RTI is able to unblock the waiting federate with a message, potentially after very little elapsed time, whereas in decentralized coordination the federate has to wait.